plugins/discourse-calendar/lib/discourse_post_event/event_finder.rb

```diff
@@ -94,39 +85,86 @@ def self.apply_privacy_restrictions(events, user)
-      return events if params[:after].blank?
-
-      after_date = params[:after].to_datetime
-      events.where(
-        "(discourse_post_event_events.recurrence_until IS NULL OR " \
-          "discourse_post_event_events.recurrence_until >= ?)",
-        after_date,
-      )
-    end
-
-    def self.apply_end_date_filter(events, params)
-      return events if params[:end_date].blank?
-
-      end_date = params[:end_date].to_datetime
-      events.where("discourse_post_event_events.original_starts_at <= ?", end_date)
-    end
-
-    def self.apply_category_filters(events, params)
+    def self.filter_by_dates(events, params)
+      return events if params[:before].blank? && params[:after].blank? && params[:end_date].blank?
+
+      before_date = params[:before]&.to_datetime
+      after_date = params[:after]&.to_datetime
+      end_date = params[:end_date]&.to_datetime
+
+      # For recurring events
+      if before_date || after_date || end_date
+        recurring_conditions = []
+        recurring_values = []
+
+        if after_date
+          # For recurring events: original start date OR recurrence_until should be >= after_date
+          # This means either the event starts after the date, or it recurs until after the date
+          recurring_conditions << "(discourse_post_event_events.original_starts_at >= ? OR discourse_post_event_events.recurrence_until IS NULL OR discourse_post_event_events.recurrence_until >= ?)"
+          recurring_values += [after_date, after_date]
+        end
+
+        if before_date
+          # For recurring events: original start date should be < before_date
+          # We want events that start before the end of our range
+          recurring_conditions << "discourse_post_event_events.original_starts_at < ?"
+          recurring_values << before_date
+        end
+
+        if end_date
+          # For recurring events: start date should be before or equal to end_date
+          recurring_conditions << "discourse_post_event_events.original_starts_at <= ?"
+          recurring_values << end_date
+        end
+
+        recurring_condition =
+          (
+            if recurring_conditions.any?
+              "(discourse_post_event_events.recurrence IS NOT NULL AND (#{recurring_conditions.join(" AND ")}))"
+            else
+              "1=0"
+            end
+          )
+
+        # For non-recurring events - check event_dates
+        non_recurring_conditions = []
+        non_recurring_values = []
+
+        if after_date
+          non_recurring_conditions << "latest_event_dates.starts_at >= ?"
+          non_recurring_values << after_date
+        end
+
+        if before_date
+          non_recurring_conditions << "latest_event_dates.starts_at < ?"
+          non_recurring_values << before_date
+        end
+
+        if end_date
+          non_recurring_conditions << "latest_event_dates.starts_at <= ?"
+          non_recurring_values << end_date
+        end
```

Comment:
No need to change it really...but it makes me uneasy when we build up these huge SQL queries this way with the arrays of values and `?` rather than using the replacement words like e.g. `:end_date`. IMO it can be easy for people to screw up this order.

This example is quite well tested though so probably fine